transport/x/cherami/inbound.go

```diff
@@ -0,0 +1,204 @@
+// PrefetchCount will result in more messages being buffered locally causing high memory foot print
+
+type InboundConfig struct {
+	Destination         string
+	ConsumerGroup       string
+	Frontend            string
+	Port                int
+	PrefetchCount       int
+	CheramiTimeoutInSec int
+}
+
+// Inbound is a inbound that uses cherami as the transport
+type Inbound struct {
+	config         InboundConfig
+	consumer       cherami.Consumer
+	router         transport.Router
+	tracer         opentracing.Tracer
+	cheramiFactory CheramiFactory
+
+	once sync.LifecycleOnce
+}
+
+// NewInbound builds a new cherami inbound
+func NewInbound(config InboundConfig) *Inbound {
+	if config.PrefetchCount == 0 {
+		config.PrefetchCount = defaultPrefetchCount
+	}
+	if config.CheramiTimeoutInSec == 0 {
+		config.CheramiTimeoutInSec = defaultCheramiTimeoutInSec
+	}
+	return &Inbound{
+		config:         config,
+		tracer:         opentracing.GlobalTracer(),
+		cheramiFactory: NewCheramiFactory(),
+	}
+}
+
+// Transports returns nil for now
+func (i *Inbound) Transports() []transport.Transport {
+	return nil
+}
+
+// SetRouter configures a router to handle incoming requests.
+// This satisfies the transport.Inbound interface, and would be called
+// by a dispatcher when it starts.
+func (i *Inbound) SetRouter(router transport.Router) {
+	i.router = router
+}
+
+// IsRunning returns whether the inbound is still processing requests.
+func (i *Inbound) IsRunning() bool {
+	return i.once.IsRunning()
+}
+
+// Start starts the inbound, reading and handling messages from cherami
+func (i *Inbound) Start() error {
+	return i.once.Start(i.start)
+}
+
+func (i *Inbound) start() error {
+	if i.router == nil {
+		return errors.ErrNoRouter
+	}
+
+	var client cherami.Client
+	var err error
+	if len(i.config.Frontend) > 0 {
+		client, err = i.cheramiFactory.GetClientWithFrontEnd(i.config.Frontend, i.config.Port)
+	} else {
+		client, err = i.cheramiFactory.GetClientWithHyperbahn()
+	}
+	if err != nil {
+		return err
+	}
+
+	consumer, ch, err := i.cheramiFactory.GetConsumer(client, i.config.Destination, i.config.ConsumerGroup, i.config.PrefetchCount, i.config.CheramiTimeoutInSec)
+	if err != nil {
+		return err
+	}
+
+	i.consumer = consumer
+
+	go func() {
+		for delivery := range ch {
+			// checksum verification before accessing message payload data
+			if !delivery.VerifyChecksum() {
+				log.Printf("checksum verification failed for ack_token: %s, asking for redelivery\n", delivery.GetDeliveryToken())
+				delivery.Nack()
+				continue
+			}
+
+			msg := delivery.GetMessage()
+
+			if err = i.handleMsg(msg.Payload.Data); err == nil {
+				if err = delivery.Ack(); err != nil {
+					log.Printf("ack failed for ack_token: %s\n", delivery.GetDeliveryToken())
+				}
+			} else {
+				if err = delivery.Nack(); err != nil {
+					log.Printf("nack failed for ack_token: %s\n", delivery.GetDeliveryToken())
```

Comment:
In this case, you're losing information about the actual handler error. We would want to log that too.

Consider using `internal/errors.CombineErrors` instead:

    // import "go.uber.org/yarpc/internal/errors"
    
    if err := i.handleMsg(..); err == nil {
        // ...
    } else {
        err = errors.CombineErrors(err, delivery.Nack())
        log.Printf(err)
    }